Sig fig improvements

pstaabp opened the pull request:
This is a branch set up to better communicate changes to the sig fig macro.

pstaabp committed:
More improvements to handle rounding better.

pstaabp committed:
More improvements to handle rounding better.

u

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -84,32 +83,28 @@ sub new {
-			$digits =~ s/[Ee].*$//;     # remove exponent, if any
+			$digits =~ s/[Ee].*$//;    # remove exponent, if any
 			if ($value == 0) {
 				$digits =~ s/^[+-]?0+(.(?:\.|$))/$1/;    # remove leading 0s before the decimal
```
I realized that there needs to be one more line after line 109:

``` perl
				$digits =~ s/^\./0./g;    # add initial zero, if needed
```

so that something like `Compute(".00")` will end up being `0.00` not `0.0` (two digits).

pstaabp replied:
I'm testing 0.00, which should have 2 sig figs.  Is line 109 supposed to replace this will .00 ?  It isn't doing that. 

And then why do we want to add another leading 0?  That ends up have an extra sig fig, doesn't it?

pstaabp replied:
I'm thinking this this block is only for numbers that are equal to 0, then 1) if the result is 0, it should be 0 with infinite sigfigs or 2) has the form 0.000  and in the 2nd case would `s/^[+-]?0+\.(0+)$/$1/` give the correct number of  sigfigs?

dpvc replied:
> I'm testing 0.00, which should have 2 sig figs.

No, this has 3 digits, not two.  The initial 0 must also be counted in order to get this to be `0.00` and not `0.0`.  That was the point of the added line.  If you have this as 2 digits, then `format()` will use `sprintf("%.1E", 0)`, which will produce `0.0` not the correct number of decimal places.

> Is line 109 supposed to replace this will .00 ?

No, it should follow 109.

> And then why do we want to add another leading 0? That ends up have an extra sig fig, doesn't it?

No, the significant digit value is really the number of digits to show, which in this case should be 3, not 2.  Just as `1.00` has three significant digits, so does `0.00`.  If you want to count the latter as two, then you have to special case 0 all over the place in the formatting and rounding functions, which makes for unnecessary complications.

dpvc replied:
> in the 2nd case would `s/^[+-]?0+\.(0+)$/$1/` give the correct number of sigfigs?

No, this will be one short, in general, and since it requires there to be a leading 0, it won't handle things like `+.00`.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -118,15 +113,15 @@ sub new {
 			}
 		}
-		$n = length($digits) unless defined($n) && $n =~ /^inf$/;    # what remains are the significant digits
+		$n = length($digits);                            # what remains are the significant digits
```
I would have the comment indented like the one on line 107, since it is at the same nesting depth as that, not the ones above.  But perhaps perltidy won't allow that.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -118,15 +113,15 @@ sub new {
-	$self->{exp}  = $n == 'inf' ? 0 : (split(/E/, $value))[1] + 0;
+	$self            = bless $self->SUPER::new($context, ROUND($value, $n)), $class;
+	$value           = $class->format('E', $self->value, $n) unless $n == 'inf';
+	$self->{data}    = [$value];
```
I still object to storing a string value rather than the perl real that a Real object is suppose to hold.  Every one of the uses of `$self->value` in this file uses it as a *numeric* value, not a string.  The only exception is line 123, which should be `$value` rather than `$self->value` anyway.  There is no point in storing the string rather than the number, as it is going to be converted to the perl real every time it is used.  That is inefficient and wasteful.  Plus it is a potential source of error, as a string sometimes acts differently from a number.

You said in you email

> I was thinking it would be better to handle the cases like you pointed out that a decimal cannot be represented as a floating point without trailing 9s.

Storing the string version is not going to make any difference to that.  When you use the value as a number (as you do in every case where you use `$self->value`, it is converted back into the internal binary representation.  Any other version of the same string representation will become the same binary number, so they will both have the exact same error and so that will not be a problem.  you are not going to get around the fact that the the binary representation is not exact.  But fortunately that means you can't compare it to an exact value.

For example, even if you store .005 as `5E-3`, if you compare that to .005, the string will be converted to the same internal perl real as .005 itself, so there is no loss in storing the .005 instead and not having to do the conversion from the string.

Unless you have a good example of where it helps, I really think you need to leave it be a real, as a Real object expects it to be.

pstaabp replied:
I think I misunderstood what was happening.  I understand our conversation to keep these as reals instead of strings.  I was thinking that we needed to store the number in some exponential form, but that's not possible.   I think the current form has fixed this.

dpvc replied:
> I think the current form has fixed this.

If you mean the lines listed above (120 through 122), then no, they haven't fixed it.  Line 122 sets the data to be the string produced in line 121, and later, line 137 does the same.  Line 120 already sets `$self->{data}[0]` to be the real value returned by `ROUND()`, so there is no need to set `$self->{data}[0]` yourself if you are going to use the perl real rather than a string.

> I was thinking that we needed to store the number in some exponential form

A perl real is already an exponential form internally.  The IEEE floating point representation uses an exponential representation, where 53 bits are used for the mantissa and 11 for the exponent, if I recall correctly.  That is why you can get values like `1.23E+50`, which would require more than 64 bits as a binary integer.  It is also why `1.23E+50 + 1` remains just `1.23E+50`, because the 1 falls below the size of the mantissa, so doesn't affect it.  That is, the real only has 16 or 17 significant digits, and so `1` is far below the least significant digit in `1.23E+50`.

pstaabp replied:
Yes.  I see the data type that format and ROUND now returns.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -118,15 +113,15 @@ sub new {
-	$self->{data} = [$value];
-	$self->{exp}  = $n == 'inf' ? 0 : (split(/E/, $value))[1] + 0;
+	$self            = bless $self->SUPER::new($context, ROUND($value, $n)), $class;
+	$value           = $class->format('E', $self->value, $n) unless $n == 'inf';
```
The `$n`s here should be `$N`, as I mentioned in my email.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -118,15 +113,15 @@ sub new {
+	$self            = bless $self->SUPER::new($context, ROUND($value, $n)), $class;
+	$value           = $class->format('E', $self->value, $n) unless $n == 'inf';
+	$self->{data}    = [$value];
+	$self->{exp}     = $n == 'inf' ? 0 : (split(/E/, $self->value))[1] + 0;
```
Same here.  Though `$n` will work, it is more consistent to use `$N`.  The only place `$n` is needed is in the `ROUND()` call at line 120.

Also, `$self->value` can just he `$value`, as these are both the same in your code, and if you take my advice about storing the perl real rather than a string, it will *have* to be `$value` here.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,72 +131,62 @@ sub new {
+	my ($self, $n) = @_;
+	return $self->{sigfigs} if !defined($n) || $self->{sigfigs} == $n;
+	$self->{sigfigs} = $self->context->checkSigFigs($n);
+	$self->{data}[0] = $n == 'inf' ? $self->value : $self->format('E', ROUND($self->value, $n), $n);
```
When you stop storing the string value and store the perl one instead, this should be

``` perl
	$self->{data}[0] = ROUND($self->value, $n);
```

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,72 +131,62 @@ sub new {
+# Return the exponential for the given $value with max(3,14-$exp) sigfigs.  This is
+# helpful for addition and subtraction.
+sub expFor {
+	my ($self, $value, $exp) = @_;
+	$exp = main::max(3, 14 - $exp);
+	return (split(/E/, $self->format('E', $value, $exp)))[1] + 0;
```
I suppose it would be better to use `$n` rather than `$exp` in these lines, as that is actually what is being computed.

Also, line 151 should probably be

``` perl
	$exp = main::max($exp, 14 - $exp);
```

(or with `$n` instead of `$exp`), as it is in the `ROUND()` function.

Those are both my fault, of course.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -316,25 +323,6 @@ sub perl {
-	$self = shift;
-	return $self->SUPER::eval(@_);
-}
+sub eval { shift->SUPER::eval(@_) }
```
I don't understand this function, as it does nothing but call the original `eval()`, so removing it would only make things more efficient.

pstaabp replied:
This was from something else that I used before.  Not sure why I kept it.

dpvc reviewed (changes requested):
You are making progress, but I still have a few items to suggest.

pstaabp committed:
changed the internal value from a string to a real and other tweaks.

pstaabp commented:
@dpvc Have a new version up.  As I mentioned above, I now understand that I was inadvertently storing a string in $self->value.  I believe this is changed now. 

Also added a expFormat (not sure this is the best name) to return the string that I had been storing to be able to do testing.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -101,26 +101,29 @@ sub new {
 			}
 		}
-		$n = length($digits);                            # what remains are the significant digits
+		$n = $digits eq '0' ? 'inf' : length($digits);     # what remains are the significant digits
```
I've discussed these changes  already.  I see that you want to have `0` have in infinite precision (even though other integers like `1` don't).  I think that can be accomplished a little easier.

``` perl
		my $digits = $value;
		$digits =~ s/^[-+]//;        # remove leading sign
		if ($value !~ m/[.Ee]/) {       # The number is an integer.
			$digits =~ s/0+$//;
		} else {
			$digits =~ s/[Ee].*$//;     # remove exponent, if any
			if ($value == 0) {
				$digits =~ s/^0*\.?/0/;    # remove leading 0s, leaving only one
			} else {
				$digits =~ s/\.//g;        # remove decimal point
				$digits =~ s/^0+//;        # remove leading zeros
			}
		}
		$n = length($digits) || 'inf';  # what remains are the significant digits
```

Here I've moved the removal of initial plus or minus signs to before the if-then conditions, since we always want to do that (and kudos for noticing that needed to be done for your 0 case, though you need to remove a minus sign as well to handle `-0`).

The non-integer zero case can now be simplified to just one substitution, which removes all leading zeros (if any) and the decimal point and replaces them with a single zero (the one that should have been before the decimal in the first place).

Finally, the setting for `$n` at line 120 can be simplified, since `length($digits)` will be zero only when `$value` has no decimal or exponential, and all the digits are zero (i.e., it is an integer 0), so `$n` will become `'inf'` in that case.  All other cases will be guaranteed to have at least one digit remaining, so you that number of digits, not `'inf'`.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -101,26 +101,29 @@ sub new {
-	$value           = $class->format('E', $self->value, $n) unless $n == 'inf';
-	$self->{data}    = [$value];
-	$self->{exp}     = $n == 'inf' ? 0 : (split(/E/, $self->value))[1] + 0;
+	$self            = bless $self->SUPER::new($context, ROUND($value, $N)), $class;
```
As I mentioned in my previous comments, this `$N` needs to remain `$n`, so that `Real(.005, sigfigs => 0)` will round up to `.01` rather than giving you `.005` with one significant figure.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -148,8 +151,8 @@ sub sigfigs {
 # helpful for addition and subtraction.
 sub expFor {
 	my ($self, $value, $exp) = @_;
```
You missed the `$exp` here.

pstaabp replied:
I'm guessing that the intention was just to rename $exp to $n.

dpvc replied:
Yes, that's correct.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -148,8 +151,8 @@ sub sigfigs {
 	my ($self, $value, $exp) = @_;
-	$exp = main::max(3, 14 - $exp);
-	return (split(/E/, $self->format('E', $value, $exp)))[1] + 0;
+	my $n = main::max(3, 14 - $exp);
```
You missed the `$exp` at the end, here.  I also suggested changing the 3 to `$n`, so that you always get at least the original precision.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -162,6 +165,11 @@ sub TeX {
 	return "{$tex}";
 }
 
+sub expFormat {
+	my $self = shift;
+	return $self->format('E', $self->value, $self->N);
+}
```
Note that the `$self->format('E')` is equivalent to `$self->format('E', $self->value, $self->N)`, so this new method is basically redundant, as `$self->format('E')` is no harder to do that `$self->expFormat`.  I'm not sure there is a need for this method, especially since your only use is in the test files.

pstaabp replied:
I'm not getting this default.  I think this line in format:
```
$n     = $self->N - $self->E unless defined $n;
```
will set $n different than `$self->N`.  I think $n is the place for rounding, not the number of sig figs, which is `$self->N`.

pstaabp replied:
Actually $n is not the place to round (it turns into that at the end as `$n-1`), however I'm not getting the same results between `$self->format('E')` and `$self->format('E', $self->value, $self->N)` I think because of the line I sited above.

dpvc replied:
Ok, you are right, the default for `$n` isn't correct for `E` notation, but is for `f` notation.  For `E` notation, one wants `$n` to be the total number of digits to use (the `$self->N` value), while for `f` notation, it should be the number of decimal places (the `$self->N - $self->E` value).

Since the only place the defaults are used is in the `string()` method, which uses `f`, I never noticed that problem.

One solution would be to change line 182 to 

``` perl
	$n = $self->N - ($f eq 'f' ? $self->E : 0) unless defined $n;
```

which gets the value right in both cases.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -174,7 +182,7 @@ sub format {
 	return "$value" if $n == 'inf';
 	$f = 'E', $n = $self->N if $f eq 'f' && ($n < 1 || $self->E >= 14 || -5 >= $self->E);
-	$n = main::max(0, $n - 1);
+	$n = $value == 0 ? $n : main::max(0, $n - 1);
```
If you make the consistency use of `$self->N` for zero (where `0.00` has three figures, as it certainly does), then you don't need this special case.

If you make `0.00` and `.00` have only two digits, then you will have to adjust the addition and subtraction functions to compensate for that, as currently `Compute("1.23 - 0.00")` will produce `1.2` rather than `1.23`, since the zero claims to have only two digits of precision.  In your version, both `0.00` and `.00` act as `0.0`, and they stringily to `0.0` rather than `0.00`.  So if you go with 2 digits, there are several other places you need to compensate for 0, and it is much easier (and less error prone) to use this as the 3 digits that you actually need rather than 2.

pstaabp replied:
I'm checking with my colleague who I'm trying to this working for.  I can definitely see that it is easier computationally if 0.00 is 3 sigfigs.     
An example that I came up with is  10.35123 - 10.35 = 0.00.  The important thing is that the hundredths is represented in the result, which I why I thought that 2 sig figs is correct.  Wondering what the chemists think about this case.

pstaabp replied:
I just heard back and I think 3 sigfigs for 0.00 seems like a good way to go--will include that in the documentation.

dpvc replied:
> An example that I came up with is 10.35123 - 10.35 = 0.00. The important thing is that the hundredths is represented in the result, which I why I thought that 2 sig figs is correct.

Of course, the number of decimal places is not the same as the number of significant figures.  The 10.35 has two decimal places, but that is not its number of significant figures.  Because 0.00 requires 3 digits to represent it makes it 3 digits.

I'm glad your colleague is OK with 3 for this.

dpvc reviewed (changes requested):
See my comments on the new commit.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -210,60 +203,77 @@ sub add {
+
+sub ROUND {
+	my ($x, $n) = @_;
+	return $x if $n == 'inf';                            # keep the same if infinite digits
```
This should probably be

``` perl
	return $x + 0 if $n == 'inf';                        # keep the same if infinite digits
```
in order to make sure that if $x is a string representing a number it gets converted to a perl real instead of staying a string.  That way, `Real('000')` will store the number 0 rather than the string `'000'`.

pstaabp committed:
Continued improvements.  Fixes to code and update to POD

pstaabp commented:
@dpvc I have another version.  

I changed format a bit to 

1. To make it handle either calling as a method of a Real or a standalone function.  When testing to format a given number/string, it didn't seem necessary to have a number created already. 
2.  Also, it seems like the last step in calculating $n, I think with our discussion of the sigfigs with 0, then we always needed to shift $n by 1. 

Also make an `expon` function to calculate the exponent of a number.  It was used a couple of times, so tried to combine this. 

Lastly, updated the tests and the POD.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -64,16 +98,45 @@ =head3 Setting the number of significant figures.
+
+	sub SigFigNumber {
+		my ($x, @opts) = @_;
+		return context::SignificantFigures::Real->new($x, sigfigs => (@opts == 1 ? $opts[0] : $opts[1]));
```
This makes the assumption that if `@opts == 2` then `$opts[0]` is `sigfigs`.  I think you might be better off using

``` perl
@opts = (sigfigs => $opts[0]) if @opts == 1;
return context::SignificantFigures::Real->new($x, @opts);
```

And for `context::SignificantFigures::Real` you shoud probably use `Context()->Package('Real')` so that if someone subclasses your context, they don't have to redefine `SigFigNumber()` (which would cause a log error).  Especially if you plan to make your context be a context extension in the end.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -85,48 +148,41 @@ sub new {
 			} else {
-				$digits =~ s/[-+.]//g;                   # remove non-digits
-				$digits =~ s/^0+//;                      # remove leading zeros
+				$digits =~ s/\.//g;        # remove non-digits
```
OOPS, I should have removed the `g` option, here, as there is only one decimal point in a number.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,134 +192,150 @@ sub new {
+	my ($self, $n) = @_;
+	return $self->{sigfigs} if !defined($n) || $self->{sigfigs} == $n;
+	$self->{sigfigs} = $self->context->checkSigFigs($n);
+	$self->{data}[0] = ROUND($self->value, $n) + 0;
```
You don't need the `+ 0` here, since `ROUND()` now always returns a perl real.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,134 +192,150 @@ sub new {
-	return sprintf("%.${m}E", $value + 0);
+# Determine the exponent of the given number.  Works like log10, but handles 0 and negatives.
+
+sub expon {
+	my $value = sprintf("%E", shift);
+	return ((split(/E/, $value))[1]) + 0;
```
There is a problem with this.  Because the default number of decimal digits for the mantissa in `E` format is 6, that means `expon(.9999999999)` will be 0 instead of -1.  Same for `expon(.999999951)`.  You might be able to get away with using `%.14E` as the format, since we are limiting to 15 significant digits.  But in that case, `$self->format($value, 0)` does the same thing, so there is already a way to do that without much trouble.  So I'm not sure this is really needed especially if it is only for use in the test file.

If you end up using `expFor($value, 0)` in place of `expon($value)`, then you should use

``` perl
sub expFor {
	my ($self, $value, $n) = @_;
	$n = main::max(0, $n - 1, 14 - $n);
	return (split(/E/, sprintf("%.${n}E", $value)))[1] + 0;
}
```

for `expFor` since `format()` will be calling `expFor()` in the direct-call use case.  That's more efficient in any case.

pstaabp replied:
I see that I was duplicating expFor, except it didn't work.  I was a little vague on what it was doing, but now understand that this is handling those edge cases.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,134 +192,150 @@ sub new {
 sub TeX {
-	my $self = shift;
-	my $tex  = $self->string;
+	my $tex = shift->string;
```
Does perltidy approve of the unaligned equals?

pstaabp replied:
I think you're thinking of lining up the = on two versions of the code.

dpvc replied:
OPPS!  Right you are.  Sorry for the noise.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,134 +192,150 @@ sub new {
+
+sub format {
+	my $self = shift;
+	my ($f, $value, $n);
+	($self, $f, $value, $n) = ref($self) eq 'context::SignificantFigures::Real' ? ($self, @_) : (undef, $self, @_);
```
This isn't really the right way to handle being able to call this function without an instance of the class.  The usual way to do that is via

``` perl
context::SignificantFigures::Real->format(...)
```

not

``` perl
context::SignificantFigures::Real::format(...)
```

That way, you always have a `$self`, but in the direct call case, it is the string `context::SignificantFigures::Real` rather than in instance of it.

With that in mind, I think you want to do something more like

``` perl
sub format {
	my ($self, $f, $value, $n) = @_;
	$value = $self->value unless defined $value;
	$n     = $self->N     unless defined $n;
	return "$value" if $n == 'inf';
	my $exp = $self->E // $self->expFor($value, 0);
	$f = 'E' if $f eq 'f' && ($n < 1 || $exp >= 5 || -5 >= $exp);
	$n -= $exp if $f eq 'f';
	$n = main::max(0, $n - 1);
	$value = ROUND($value, 0) if $n == 0;
	return sprintf("%.${n}${f}" . ($n == 0 && $f eq 'f' ? '.' : ''), $value);
}
```

When called directly, `$self->E` will return `undef`, and so 

``` perl
my $exp = $self->E // $self->expFor($value, 0);
```

will end up using `$self->expFor($value, 0)` to get `$exp`.

I've also added in the use of `ROUND()` when `$n` is zero, so that `format(0.005, 0)`  will give .01 rather than .005 as it currently does.

pstaabp replied:
Yes.  I knew it wasn't the best way to handle it, but I see now.  

In terms of the 0 sigfigs, it seems like if the leftmost sig digit is >=5, it rounds up to the next power of ten, if less than 5 , to zero. 

I'm guessing this is getting the cases like 12.01-12.005=0.00 to be correct.?  Is that a good way to think about it? 

I keep throwing in more test cases (like this) to help me understand this a bit better.

dpvc replied:
> In terms of the 0 sigfigs, it seems like if the leftmost sig digit is >=5, it rounds up to the next power of ten, if less than 5 , to zero.

Right.

> I'm guessing this is getting the cases like 12.01-12.005=0.00 to be correct.?

No, because none for he terms or results has 0 digits.  But for something like 12.01 - .005, where the .005 must be rounded to have only 2 decimal places.  That is, it has 0 of its original significant digits, and must become .01.

Because the `add` and `sub` methods round first and then add or subtract, the `ROUND()` that I added here won't come into play for those, as you will never have anything but zeros below the needed precision (so no 5 to round up).  But now that you are allowed to call `format()` directly (as you are in the tests), you need to handle the case of something like `context::SignificantFigures::Real->format(.005, 0)`.

> I keep throwing in more test cases (like this) to help me understand this a bit better.

Yes, you should always try to push the limits and test the edge cases.  Try things out with 0, or with very small and very large values, or ones with different magnitudes, etc.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,134 +192,150 @@ sub new {
-	return $n <= 0 ? $self->new(0) : $self->new($value, sigfigs => $n);
+	my $exp   = main::min($l->N - $l->E, $r->N - $r->E);
+	my $value = $l->round($exp) + $r->round($exp);
+	return $self->new($value, sigfigs => $exp + $self->expFor($value, $exp));
```
In doing some more testing, it looks like you need to restrict the `sigfigs` using

``` perl
	return $self->new($value, sigfigs => main::max(0, $exp + $self->expFor($value, $exp)));
```

otherwise `Real(-100.005 + 100)` will report an error about negative significant figures.

dpvc commented on macros/contexts/contextSignificantFigures.pl:
```diff
@@ -136,134 +192,150 @@ sub new {
-	return $self->new("$v");
+	my $exp   = main::min($l->N - $l->E, $r->N - $r->E);
+	my $value = $l->round($exp) - $r->round($exp);
+	return $self->new($value, sigfigs => $exp + $self->expFor($value, $exp));
```
Similarly, you need

``` perl
	return $self->new($value, sigfigs => main::max(0, $exp + $self->expFor($value, $exp)));
```

here as well

dpvc reviewed (changes requested):
I haven't gone over the comment changes, but have some comments on the code.

dpvc commented on t/contexts/significant_figures.t:
```diff
@@ -23,254 +23,364 @@ use Data::Dumper;
 
+sub ROUND  {&context::SignificantFigures::Real::ROUND}
+sub expon  {&context::SignificantFigures::Real::expon}
+sub FORMAT {&context::SignificantFigures::Real::format}
```
With my recommended changes, this would be

``` perl
sub expon { context::SignificantFigures::Real->expFor($_[0], 0) }
sub FORMAT { context::SignificantFigures::Real->format(@_) }
```

pstaabp committed:
Adding more tests and some additional tweaks to the context.

pstaabp commented:
@dpvc  Just pushed another version. 

I was getting a warning about non-numeric equality check in the sigfigs subroutine when passing non-numeric values (I was checking for those).  I rewrote it to do the check before checking the inequality.  I could filter out those on line 197, but rewrote it a bit.  

Also, If a number like 0.8 was parsed, it has 1 sig fig and the end of the format subroutine rounds to $n-1 which is 0.  I this line 241 needs to be above the `$n = main:max($n-1, 0);`

dpvc commented:
> I [think] this line 241 needs to be above the `$n = main:max($n-1, 0);`

You are right it needs to be earlier than I had it.  I think it really needs to go earlier yet, as line 240 changes `$n` from being the number of digits of precision to the number of decimal digits when `f` format is used.  So it needs to go before that.  I'd actually move it up to  between lines 237 and 238.

I still think the `SigFigNumber` should use `Value->Package("Real()")` rather than an explicit `context::SignificantFigures::Real`, but that is minor for now.

Otherwise, I think it is in good shape.  Yay!

pstaabp committed:
A couple of minors fixes.

pstaabp commented:
I made those changes.  Thanks for getting this in good shape. 



Here's some other thoughts with this:

1. We'd like to have students be able to enter numbers in scientific notation and then parse correctly, like '1.0 x 10^2' for 100 with two sig figs, instead of requiring them to enter '1.0E+02'  It seems like we would need to parse this before `->new` is called.  How would I try to accomplish that? 
2. I'm planning to do the `Limited` version context.  I think I can handle that. 
3. We'd eventually like to role the Units context into this with the new extension macro.   Thoughts on how this might happened? 

By the way, I mentioned other functions to Steven Fiedler, who I'm working with on this.  He looked into it and sounds like it is quite complicated.  For example, because of the decimal basis of significant figures, you can do 10**x and log10, but the natural exponential and log is much more complicated.  I think power should be straightforward since it is repeated multiplication--not sure about square root though. 

I might go ahead and merge this and then start a fresh PR and then roll these other features in step-by-step.

pstaabp commented:
About #1, I found the `cmp_preprocess` which should allow me do to this.

dpvc commented:
> About https://github.com/pstaabp/pg/pull/1, I found the `cmp_preprocess` which should allow me do to this.

Note that that is only called when processing student answers, so won't be able to be used to write correct answers.  If you want to allow it in both student and correct answers (i.e., have `Compute()` be able to process it), then you need to make it known to the parser.  There are several alternatives for how to do this, each with pros and cons.

One would be to change the pattern for numbers to allow the `x 10^` notation in addition to `E`, and have the `new()` method turn that into `E` before further processing.  This would allow `1.23 x 10^5` to be treated as a single unit.  It is pretty easy to implement this approach, but it does have some drawbacks.  First, because this is treated as a unit, the `x` and `^` don't act like operators usually do.  For example, you wouldn't be able to do `1.23 x 10^(2+3)`, and `1.23 x 10^5^2` would be treated as `(1.23 x 10^5)^2` rather than the way the precedences would have it treated, as `1.23 x 10^(5^2)`.  you probably don't care about that, however.  You also don't get an error for something like `1.23 x 10^3 x 10^2`, for example, as that will be treated as `1.23E3*x*100`, where `x` is a variable (once you implement Formulas in this context).  So while this is an attractive approach, it does have some unfortunate consequences.

A second approach would be to define `x` as an operator and have it build the number from its two operands.  This would take a little care in checking the operands to make sure they are the right things, but is not too hard to do.  You would need to make `^` be available and handle its operands as well.  With some care, this can be made to work pretty well.  The down side here is that you can't then use `x` as a variable, so when you add support for Formulas in this context, you won't be able to use functions of `x` since `x` will already mean "times".  (This is why the cross product in Vector context is given by `><` not `x`.)  It would be possible to define U+00D7 (×) as the multiplication sign, leaving `x` free to be used as a variable, but that would require students to enter unicode characters, which they may not know how to do.  You could give them this character in the output and have them copy and past it, to make it easier, but it is still a little out of the ordinary, and some students may have difficulty with that.

A third way would be to make `x 10^` be a multi-character operator (along with several aliases like `x 10**` and `x10^`).  This would allow you to use `x` as a variable but still be able to write `1.23 x 10^5`.  You would be able to give appropriate error for things like `1.23 x 10^3 x 10^2`, but again, you might get things like `1.23 x 10^3 x 10` being `1.23E3*x*10`.  One would also need to do a little work to get parentheses when needed in some situations, but it can be done.  This might be the best alternative if you don't want to have to enter unicode characters.

For (2), I will let you handle that.

For (3), since the Units class is an extension class itself, it can be applied to the SignificantFigures class.  There is an issue that must be addressed however, which is that, since the values are multiplied by the various conversion factors, the precision of those factors comes into play.  Since many have a factor of 1, that means everything goes to one digit of precision.  Not good.  There are several possible solutions to that, and I will write them up later.  I am out of time at the moment.

Decide which of the three approaches above you think might be best for (1), and I can walk you through that.

dpvc commented:
In terms of (3) above, you can use 

``` perl
Context(context::Units::extends("SignificantFigures")->withUnitsFor(...))
```

to get a SignificantFigures context with the desired units.  The problem is that when some computations involving units are performed, the units are multiplied by the appropriate factor to be able to combined them.  For example, for addition, the second operand is multiplied by the needed factor to bring it to the same units as the first, and then they are added.  For example, `5 cm + 3.62 m` will multiply the 3.63 by 100 to get `367 cm`.  But when you do that within the SignificantFigures context, the 100 has only one significant digit, so 3.63 * 100 becomes 400 and the sum is just 4E+02 as a result.  Even when the two have the same units, they are multiplied by the factor 1, which has only one digit, so `1.57 m + 3.62 m` becomes `1.57 m + 4 m` or `5 m`.  Neither of these results is a good outcome.

One possible solution would be to loop through the units that have been added to the context and convert their factors to to SignificantFigures reals with infinite precision (so that the products with the factors retain the precision of the original number).

These examples, however, bring up an issue that I hadn't really considered earlier:  if a SignificantFigure number is used in a computation with a perl real somewhere inside a WeBWorK macro or in a method of a MathObject, then the perl real will be promoted to a `context::SignificantFigures::Real` and it will get the number of digits used in the perl real.  For example:

``` perl
sub ave {
  my $sum = 0;
  $sum += $_ for (@_);
  return $sum / @_;
}

Context("SignificantFigures");

TEXT(ave(3.11, 10.49, 6.72));  # produces 6.77333333333333
TEXT(ave(Real(3.11), Real(10.49), Real(6.72))); # produces 7
```

The second `ave()` call produces a number with only 1 significant digit because the division by `@_`, which is 3 in this case, has only one digit.  This case cause be corrected by using `Real(@_)`, but if the `ave()` macro is hidden inside some standard macro or object method, this can easily lead to unexpected and difficult-to-locate bugs, like the ones introduced by using the Units context.

A possible solution is to change how perl reals are promoted during computations.  Currently, the `checkOpOrderWithPromote()` calls in the `add`, `sub`, `mult`, and `div` methods is where the promotion to SignificantFigures reals is being done.  The parent class calls `$self->promote($op)` on the other operand to make sure it is the same class as the SignificantFigure real (whichever operand it is).  The default `promote()` method calls `$self->new()` to turn the other operator into a SignificantFigure real, and that restricts the number of digits as in the examples above.

We can override the `promote()` method with a new one, however, that turns a perl real into a SignificantFigures real, but with infinite precision, so that computations like he ones above will operate as they are expected to even when they apply to a SignificantFigure real.

Here is a version that does that:

``` perl
sub promote {
	my $self = shift;
	my $context = (Value::isContext($_[0]) ? shift : $self->context);
	my $value   = (scalar(@_)              ? shift : $self);
	return $value->inContext($context) if Value::isValue($value) && $value->{sigfigs};
	return $self->new($context, $value, sigfigs => 'inf');
}
```

The `promote()` call may have a context as the first argument, so we deal with that.

With that, the second `ave()` call above returns 6.773, as expected (since the sum of the three terms is 10.32, with 4 digits, so when divided by 3 with infinite digits, you get a 4-digit result).

This solution resolves the units problem (since the factors are perl reals) without having to do anything special about units, so that is nice.  And it should avoid problems with using SignificantFigures reals with constants that are within other macro calls without having to modify them.  So I think this is probably the right way to go.

dpvc commented:
Given a number $a$ with $`a_n`$ digits, we can write $`a = a_m \times 10^{a_e}`$ where $`a_m`$ has $`a_n`$ digits, and $1 \le a_m < 10$.  Then if $a$ represents the actual value $\hat a$, and $`\epsilon = 5 * 10^{a_e - a_n}`$, then $a - \epsilon \le \hat a < + \epsilon$.

dpvc commented:
Sorry!  I am writing something up about other functions, and hit comment when I didn't mean to!

dpvc commented:
You bring up the issue of handing functions like natural logs.  I mentioned in an earlier email an idea that I had for how to handle that case.  I describe it in more detail here.

Given a number $a$ with $a_n$ digits, we can write $a = a_m \times 10^{a_e}$ where $a_m$ has $a_n$ digits, and $1 \le a_m < 10$.  Then if $a$ represents the actual value $\tilde a$ to $a_n$ digits and $\epsilon_a = 5 \times 10^{a_e - a_n}$, then $a - \epsilon_a \le \tilde a < a + \epsilon_a$.  That is, the first digit following $a_m$ in $\tilde a$ would either have caused $\tilde a$ to round up to $a_m$ (the first inequality) or round down to $a_m$ (the second inequality).

Conversely, if we have an approximation $a$ for the value $\tilde a$ and know that $a$ is within $\epsilon_a$ of $\tilde a$, then we can ask how many digits of $a$ are significant?

To answer this, we must find the smallest integer $\delta_a$ where $\epsilon_a < 5 \times 10^{\delta_a}$.  Then $10^{\delta_a}$ gives the decimal place where the error in estimating $\tilde a$ by $a$ is contained in the interval that would round $a$ to the decimal place just above $10^{\delta_a}$.  That is, we should use $a_m$ having $a_n$ digits where $a_e - a_n = \delta_a$, or $a_n = a_e - \delta_a$.

To find $\delta_a$ we solve $\epsilon_a < 5 \times 10^{\delta_a}$ by writing $\epsilon_a / 5 < 10^{\delta_a}$, and then (since $\log_{10}$ is an increasing function, so preserves inequalities), we get $\log_{10}(\epsilon_a / 5) < \delta_a$.  Since we want $\delta_a$ to be the smallest integer that satisfies this, we take $\delta_a = \lceil\log_{10}(\epsilon_a / 5)\rceil$, where $\lceil \cdot \rceil$ is the ceiling function.  So $a_n = a_e - \lceil\log_{10}(\epsilon_a / 5)\rceil$ gives the number of reliable digits in an estimation $a$ of $\tilde a$ that has a known maximum error of $\epsilon_a$.

We can use this idea to determine the number of significant digits to retain when computing the value of a function given an argument with a given number of significant digits.

Suppose $f(x)$ is a function.  If we have an approximation $a$ to a value $\tilde a$ with $a_n$ significant digits, we can use $f(a)$ to approximate $f(\tilde a)$ and would like to know how many significant digits there are in the result.  I propose doing this using the linear approximation to $y=f(x)$ at $x=a$ given by $y = \hat f(x) = f(a) + (x - a)f'(x)$.  If we let $\epsilon_a = 5 \times 10^{\delta_a}$ as above, where $\delta_a = a_e - a_n$ and $a_e$ is as described above, then since $\tilde a$ is between $a-5\times 10^{\delta_a}$ and $a+5\times 10^{\delta_a}$ and $\hat f$ is linear, $\hat f(\tilde a)$ is between $\hat f(a-5\times 10^{\delta_a})$ and $\hat f(a+5\times 10^{\delta_a})$.  Plugging these values into the linear formula for $\hat f$ gives $\hat f(\tilde a)$ being between $f(a) - (5 \times 10^{\delta_a})\cdot f'(a)$ and $f(a) + (5 \times 10^{\delta_a}) \cdot f'(a)$.  That is, the error in approximating $\hat f(\tilde a)$ by $f(a)$ is at most $(5\times 10^{\delta_a}) \cdot |f'(a)|$, where the absolute values are used because $f'(a)$ might be negative.  Since we are taking $\hat f(\tilde a)$ to represent $f(\tilde a)$ we take our error in using $f(a)$ for $f(\tilde a)$ to be at most $(5\times 10^{\delta_a}) \cdot |f'(a)|$.

If $f(a) = f_m \times 10^{f_e}$, we can use the formula above to compute the number $f_n$ of digits of precision (i.e., significant digits) in $f_m$ as an approximation for $f(\tilde a)$.  That is,

$$
\begin{aligned}
f_n &= f_e - \lceil\log_{10}(\epsilon_f / 5)\rceil \\
  &= f_e - \left\lceil\log_{10}\left({1\over 5}(5\times 10^{\delta_a})\cdot |f'(a)|\right)\right\rceil \\
  &= f_e - \lceil\log_{10}(10^{\delta_a}\cdot |f'(a)|)\rceil \\
  &= f_e - \lceil\log_{10}(10^{\delta_a}) + \log_{10}|f'(a)|\rceil \\
  & = f_e - \lceil \delta_a + \log_{10}|f'(a)|\rceil
\end{aligned}
$$

That is, $f_n$ determines the number of significant digits to use in $f(a)$, and we can round $f(a)$ to that number of digits.

The idea is that the slope, $f'(a)$, is the change in $y$ for a unit change in $x$, so for a change of at most $\epsilon_a$ in $a$, we should get a change of at most $\epsilon_a f'(a)$ in $f(a)$.  That tells us the maximum error in our approximation $f(a)$ of $f(\tilde a)$, and we find the smallest $\delta_f$ that encloses that error.  That tells us where the least precise digit is, and $f_e$ minus that gives us the number of digits that we can rely on.

The end result is that when the magnitude of the slope is large (i.e., the function is steep), then a small change in $a$ makes for a large change in $f(a)$, and so we have fewer reliable digits in the result, while for a small-magnitude slope, changes in $a$ make very little difference to $f(a)$, so we get more reliable digits in the result, even with fewer digits to start with.

One can apply this idea to addition and multiplication in order to recover the usual rules of significant digits (with a few caveats), which is an interesting exercise.  I may write that up just to get my reasoning down in case I need it in the future, since I have worked out the details, and will probably forget them if I don't.

dpvc commented:
# Some thoughts on significant figures

Suppose $a$ is a number with $a_n$ significant figures.  Then $a$ can be written as $a = a_m\times 10^{a_e}$ where $a_m$ has $a_n$ digits and $1 <= |a_m| < 10$.  Suppose further that $a$ approximates some more precise number $\tilde a$ to $a_n$ digits.  Then $\tilde a = a + \epsilon_a$ for some $\epsilon_a$ where $|\epsilon_a| \le 5 \times 10^{\delta_a}$ where $\delta_a = a_e - a_n$.  Here $\delta_a$ gives the power of 10 that is just below the least significant digit in $a$.

Conversely, suppose $a$ is an approximation of $\tilde a$ where we know that $|\tilde a - a| < \epsilon_a$ for some some error bound $\epsilon_a$.  Then if $\delta_a$ is the smallest integer where $\epsilon_a < 5\times 10^{\delta_a}$, then as an approximation of $\tilde a$, we can say that $a$ has $a_n = a_e - \delta_a$ significant digits.  Solving $\epsilon_a < 5\times 10^{\delta_a}$ for $\delta_a$, we get $\delta_a = \lceil\log_{10}(\epsilon_a / 5)\rceil$. so $a_n = a_e - \lceil\log_{10}(\epsilon_a / 5)\rceil$.

We can apply this idea to determine the rules of significant figures for addition and multiplication (and so for subtraction and division as well).

## Addition

Suppose $a$ is an approximation for $\tilde a$ where $a+\epsilon_a=\tilde a$.  As above, we can write $a=a_m \times 10^{a_e}$, where $a_m$ has $a_n$ digits, and $1 \le |a_m| < 10$.  Since $a$ approximates $\tilde a$ with $a_n$ digits, we can write $\epsilon_a = m_a \times 10^{\delta_a}$ where $-5 \le m_a < 5$ and $\delta_a = a_e - a_n$.   Similarly if $b$ approximates $\tilde b$ with $b+\epsilon_b = \tilde b$ with $b_n$ digits, then we have $b=b_m\times 10^{b_e}$ and $\epsilon_b = m_b\times 10^{\delta_b}$ with $-5\le m_b < 5$ and $\delta_b = b_e - b_n$.  We can assume, without loss of generality, that $\delta_a \ge \delta_b$.

Since $a$ approximates $\tilde a$ and $b$ approximates $\tilde b$, that means $a+b$ approximates $\tilde a + \tilde b$ and $\tilde a + \tilde b = (a + \epsilon_a) + (b + \epsilon_b) = (a+b) + \epsilon_a + \epsilon_b$ and we see that

$$
\begin{aligned}
\epsilon_{a+b} &= \epsilon_a + \epsilon_b \\
  &= m_a\times 10^{\delta_a} + m_b\times 10^{\delta_b} \\
  &= m_a\times 10^{\delta_a} + m_b\times 10^{\delta_a - (\delta_a - \delta_b)} \\
  &= m_a\times 10^{\delta_a} + m_b\times {10^{\delta_a} \over 10^{\delta_a - \delta_b}} \\
  &= 10^{\delta_a}\left(m_a + {m_b \over 10^{\delta_a - \delta_b}}\right).
\end{aligned}
$$

In order to determine the number $`(a+b)_n`$ of significant digits in $a+b$, we need to use the formula above, $`(a+b)_n = (a+b)_e - \delta_{a+b}`$, so we need to find $\delta_{a+b}$, the power of ten for the first non-significant digit in $a+b$.  This is obtained from $\delta_{a+b} = \lceil\log_{10}|\epsilon_{a+b} / 5|\rceil$, which gives us

$$
\begin{aligned}
\delta_{a+b} &= \left\lceil\log_{10}\left|{1\over 5}\cdot 10^{\delta_a}\left[m_a + m_b{1\over 10^{\delta_a-\delta_b}}\right]\right|\right\rceil \\
  &= \left\lceil\log_{10}\left(10^{\delta_a}\right) + \log_{10}\left|{1\over 5}\left[m_a + {m_b \over 10^{\delta_a-\delta_b}}\right]\right|\right\rceil \\
  & = \delta_a + \left\lceil\log_{10}\left({1\over 5}\left|m_a + {m_b \over 10^{\delta_a-\delta_b}}\right|\right)\right\rceil
\end{aligned}
$$

The effect of $m_b \over 10^{\delta_a- \delta_b}$ is to shift $m_b$ farther to the right, making it less significant, and the value of $m_a$ will be the one that dominates, and so determines the value of the log in the final expression above; that is $\log_{10}\left({1 \over 5} \left|m_a + {m_b\over 10^{\delta_a-\delta_b}}\right|\right)$ is effectively $\log_{10}(|m_a| / 5)$.  We know that $|m_a|\le 5$, so $|m_a| / 5 \le 1$, and $\log_{10}(|m_a|/ 5) \le 0$.  Taking the ceiling function of this, the result will be at most 0, and we can conclude that $\delta_{a+b} \le \delta_a$.  That means that the position of the first insignificant digit in $a+b$ is, in the worst case, the same as that for $a$, which is the summand that is the least precise of the two (since we assumed $\delta_a \ge \delta_b$).  This gives the usual rule for addition with significant figures:  the result has the same precision as the least precise summand.

There is a problem with this analysis, however.  The claim that $m_b\over 10^{\delta_a-\delta_b}$ can be ignored is not actually true, as that term can affect the final logarithm under the right circumstances.  The greater the value of $\delta_a - \delta_b$, the less impact $m_b$ will have, but given the proper values of $m_a$ and $m_b$, it can still change the outcome.  

For example, with $a = 35.7$ estimating $\tilde a = 35.7483$ and $b=2.14$ estimating $\tilde b = 2.14311$, we have $m_a = 4.83$ ($\epsilon_a = 4.84\times 10^{-2}$) and $m_b = 3.11$ ($\epsilon_b = 3.11\times 10^{-3}$).  So $m_a + {m_b\over 10^{\delta_a-\delta_b}}$ = $4.83 + {3.11\over 10^{-2 - (-3)}} = 4.83 + {3.11\over 10^1} = 4.83 + .311 = 5.141$, and so this is *bigger* than 5 (not less than or equal to 5, as claimed above), and so when divided by 5 it is greater than 1, and its log to base 10 is greater than 0.  That means the ceiling will make it 1, and we end up with $\delta_{a+b} = \delta_a + 1$, giving us one fewer reliable digit in the result than we had in the summands (one fewer than we determined above).

In this case, although the least precise summand had a digit in the tenths place, we can't rely on the tenth place in the sum after all, but must round to an integer instead.  That is, $a+b = 35.7 + 2.14 = 37.84$ should not be rounded to 37.8, but rather to 38.  Note that the correct sum is $\tilde a + \tilde b = 35.7483 + 2.14311 = 37.88941$, which rounds to 37.9 not 37.8 to three places.  So our result of 37.8 is not accurate to three places, but only to two.  To further support this, note that 37.8 differs from 37.88941 by $.0894 = 8.94\times 10^{-2}$; but $8.94 > 5$, so the smallest $\delta$ where $.0894 < 5 \times 10^\delta$ is $\delta = -1$ not $-2$, and we can't include the first decimal place in an accurate answer from these approximations, but must round to 38.  With that in mind, note that $38$ differs from $37.88941$ by $.116 = 1.16\times 10^{-1} < 5 \times 10^{-1}$, so 38 is the most accurate representation of $37.88941$ that we can obtain for the approximation $37.84$.

Because the effect of $m_b$ is less when $\delta_a - \delta_b$ is larger, the worst-case scenario is the common case of having the same precision for both $a$ and $b$, when $m_b$ is not shifted at all, and we are looking at $\log_{10}\left|{m_a + m_b \over 5}\right|$.  In this case, the sum $m_a + m_b$ can easily exceed 5, meaning the log will be greater than 0, and we lose a digit as we did above.  For example, with $a = 35.1$ estimating $\tilde a = 35.14$ and $b = 21.3$ estimating $\tilde b = 21.33$, we have $a+b = 35.1+21.3=56.4$, while $\tilde a + \tilde b = 35.14+21.33=56.47$. That is, the three-digit answer should be $56.5$, not $56.4$, and we are only precise to two digits, not three.  As above, a computation for $\delta_{a+b}$ will yield $-1$, not $-2$, in this case.

In practice, however, since $-5 \le m_a < 5$ and $-5 \le m_b < 5$, the errors often reduce the sum rather than increase it, and in fact, in 75% of the time $|m_a + m_b|$ will remain less than 5 (assuming every value for $m_a$ and $m_b$ is equally probable), and only 25% of the time will it exceed it and cause a loss of a digit (note also that at most one digit can be lost, as the sum would have to exceed 50 to lose two digits, but since both are less than 5, it can't exceed 10).  Given these percentages, it is reasonable to make the rule be to use the least significant precision of the two summands and ignore the case where an extra digit is lost.  Also, if the rule took that into account, you would quickly run out of digits when making a sum of several numbers!  So the addition rule as we know it is a practical one, if not a perfect one, and is well supported by these computations.


## Multiplication

As above, suppose $a$ is an approximation for $\tilde a$ where $a+\epsilon_a=\tilde a$ with $a=a_m \times 10^{a_e}$, where $a_m$ has $a_n$ digits and $1 \le |a_m| < 10$, and $\epsilon_a = m_a \times 10^{\delta_a}$ where $-5 \le m_a < 5$ and $\delta_a = a_e - a_n$.   Similarly $b$ approximates $\tilde b$ with $b+\epsilon_b = \tilde b$ with $b_n$ digits, with $b=b_m\times 10^{b_e}$ and $\epsilon_b = m_b\times 10^{\delta_b}$ with $-5\le m_b < 5$ and $\delta_b = b_e - b_n$.  We can assume, without los of generality, that $a_n \le b_n$, so $a$ has fewer significant digits than $b$.

Since $a$ approximates $\tilde a$ and $b$ approximates $\tilde b$, that means $ab$ approximates $\tilde a \tilde b$ and $\tilde a \tilde b = (a + \epsilon_a)(b + \epsilon_b) = ab + \epsilon_a b + \epsilon_b a + \epsilon_a\epsilon_b$ and we see that

$$
\begin{aligned}
\epsilon_{a+b} &= \epsilon_a b + \epsilon_b a + \epsilon_a \epsilon_b \\
  &= (m_a\times 10^{\delta_a})(b_m\times 10^{b_e}) + (m_b\times 10^{\delta_b})(a_m\times 10^{a_e}) +  (m_a\times 10^{\delta_a})(m_b\times 10^{\delta_b})\\
  &= (m_a b_m)\times 10^{\delta_a + b_e} + (m_b a_m)\times 10^{\delta_b + a_e} + (m_a m_b)\times 10^{\delta_a + \delta_b} \\
  &= (m_a b_m)\times 10^{a_e - a_n + b_e} + (m_b a_m)\times 10^{b_e - b_n + a_e} + (m_a m_b)\times 10^{a_e - a_n + b_e - b_n} \\
  &= (m_a b_m)\times 10^{a_e +b_e - a_n} + (m_b a_m)\times 10^{a_e + b_e - b_n + a_n - a_n} + (m_a m_b)\times 10^{a_e + b_e - a_n - b_n} \\
  &= 10^{a_e+b_e - a_n}\left(m_a b_m + (m_b a_m)\times 10^{a_n - b_n} + (m_a m_b)\times 10^{-b_n} \right) \\
  &= 10^{a_e+b_e - a_n}\left(m_a b_m + {m_b a_m\over 10^{b_n - a_n}} + {m_a m_b\over 10^{b_n}} \right)
\end{aligned}
$$

In order to determine the number $`(ab)_n`$ of significant digits in $ab$, we need to use the formula above, $`(ab)_n = (ab)_e - \delta_{ab}`$, so we need to find $\delta_{ab}$, the power of ten for the first non-significant digit in $ab$.  This is obtained from $\delta_{ab} = \lceil\log_{10}|\epsilon_{ab} / 5|\rceil$, which gives us

$$
\begin{aligned}
\delta_{ab} &= \left\lceil\log_{10}\left|{1\over 5}\cdot 10^{a_e+b_e - a_n}\left[m_a b_m + {m_b a_m\over 10^{b_n - a_n}} + {m_a m_b\over 10^{b_n}} \right]\right|\right\rceil \\
  &= \left\lceil\log_{10}\left(10^{a_e+b_e - a_n}\right) + \log_{10}\left|{1\over 5}\left[m_a b_m + {m_b a_m\over 10^{b_n - a_n}} + {m_a m_b\over 10^{b_n}}\right]\right|\right\rceil \\
  & = (a_e + b_e) -a_n + \left\lceil\log_{10}\left({1\over 5}\left|m_a b_m + {m_b a_m\over 10^{b_n - a_n}} + {m_a m_b\over 10^{b_n}}\right|\right)\right\rceil
\end{aligned}
$$

As in the case above, the effect of $10^{b_n-a_n}$ is to shift $m_b a_m$ to the right, and similarly, $10^{b_n}$ shifts $m_a m_b$ to the right, since $b_n \ge 1$.  That leaves $m_a b_m$ as the dominant factor, so the log in the final line above is essentially $\log_{10}(|m_a| |b_m| / 5)$.  But we know $|m_a| \le 5$ and $1 \le |b_m| < 10$, so $|m_a| |b_m| / 5 < 10$, and $\log_{10}(|m_a| |b_m| / 5) < \log_{10}(10) = 1$.  Taking the ceiling of this gives us at most 1.

This means that the worst case is $\delta_{ab} = (a_e+b_e) - a_n + 1$.  In order to find the number of significant digits, we need to determine $`(ab)_e`$, the magnitude (power of 10) of $ab$.  Since $ab = (a_m\times 10^{a_e})(b_m\times 10^{b_e})=(a_m b_m)\times 10^{a_e + b_e}$, the magnitude will be $a_e+b_e$ as long as $|a_m b_m|$ is less than 10, and $a_e + b_e + 1$ if $|a_m b_m|$ is bigger than or equal to 10 (we know $|a_m| < 10$ and $|b_m| < 10$, so $|a_m b_m| < 100$ so we will never get more than plus 1).

Since $1\le |a_m| < 10$ and $1\le |b_m| < 10$, it turns out that $|a_m b_m| \ge 10$ just under 72% of the time (assuming every value for $a_m$ and $b_m$ are equally probable), so $`(ab)_e = a_e+b_e + 1`$, and in this case, $`(ab)_n = (ab)_e - \delta_{ab} = (a_e+b_e) + 1 - [(a_e+b_e) - a_n + 1] = a_n`$.  That means, the number of significant digits in $ab$ is the same as the number is $a$, the term with the fewest significant digits.  This gives the usual rule for multiplication: the number of significant digits in a product is the same as the smallest number of significant digits in either of the factors.

In 28% if the cases, however, $|a_m b_m| < 10$, and so $`(ab)_e = a_e + b_e`$, meaning $`(ab)_n = (a_e+b_e) - [(a_e+b_e) - a_n + 1]=a_n-1`$, and we lose one digit of precision.  As above, from a practical standpoint, however, keeping the same number of digits makes sense, as we would quickly lose all our digits after repeated multiplications.

Of course, our argument here suffers the same problem as it did in the case of addition:  the effect of $m_b a_m / 10^{b_n-a_n} + m_a m_b / 10^{b_n}$ can't always be ignored, and can cause us to lose a digit of precision more often.  On the other hand, since these terms can be negative, which may lead to better precision, and since we assumed the worst case by taking $\delta_{ab} = (a_e+b_e) - a_n + 1$, we may not lose a digit even when $|a_m b_m| < 10$.  So the multiplication rule is probably better than it looks from this analysis.

As an aside, because there is a non-zero probability of losing a digit during multiplication, that probability increases with repeated multiplication.  So in computing $x^m$ by repeated multiplication for an $n$-digit $x$, the likelihood that you continue to have $n$ digits in the result decreases as $m$ increases.  I suspect that using the $f'$ approach I discussed in an earlier comment for $f(x) = x^m$ will give a better idea of the reliable significant digits than keeping $n$ digits in the result for repeated multiplication.
